File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> answer = new ArrayList<>();
        Arrays.sort(nums);
        
        for (int i = 0; i < nums.length - 2; i++) {
            if (i != 0 && nums[i] == nums[i - 1]) continue;
            int lt = i + 1, rt = nums.length - 1;
            
            while(lt < rt) {
                int sum = nums[i] + nums[lt] + nums[rt];
                
                if (sum == 0){
                    answer.add(List.of(nums[i], nums[lt], nums[rt]));
                    
                    while (lt < rt && nums[lt] == nums[lt + 1]) {
                        lt++;
                    }
                    
                    while (lt < rt && nums[rt] == nums[rt - 1]) {
                        rt--;
                    }
                    
                    rt--;
                    lt++;                        
                }
                else if (sum > 0) {
                    rt--;
                } else {
                    lt++;
                }
            }
        }
        
        return answer;
        
    }
}

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        answer = []
        
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
                
            lt, rt = i + 1, len(nums) - 1
            
            while lt < rt:
                total = nums[lt] + nums[rt] + nums[i]
                
                if total > 0:
                    rt -= 1
                elif total < 0:
                    lt += 1
                else:
                    answer.append([nums[i], nums[lt], nums[rt]])
                    
                    while lt < rt and nums[lt] == nums[lt + 1]:
                        lt += 1
                    while lt < rt and nums[rt] == nums[rt - 1]:
                        rt -= 1
                    
                    lt += 1
                    rt -= 1
        
        return answer
                        